### ISSUE TITLE
Wiping memory with a room with more than one spawn

### ISSUE THREAD

**Issue by quonic** [bonzaiferroni/bonzAI#1]
When wiping memory more than one spawn in a room results in that room not finding the centerPosition.

It looks like in the initWithSpawn() in the ControllerOperations.ts is only looking for one spawn.

There should be a way to find the center of what is already built after a memory wipe. As in the distance from the top most structure to the bottom most, or left most and right most. This would assume there isn't some random road or something inline with the spawns.

Edit: Also great code. Very elegant and readable. 👍

**Comment by bonzaiferroni** [comment]
Thanks for taking it for a spin! 

`cc.wipeMemory()` was a quickly hacked-out solution to getting rid of operation memory that was no longer being used. But since then I've added the autolayout stuff, and that should obviously be part of what is saved. Since then, I've thought about whether it would be good to have memory values that can't be easily recalculated protected in another way, but I haven't really thought of a solution. In the meantime, I put the cautionary note in the JSDoc about that, but it might be safer to comment it out.

I've also thought about how an existing layout could be reconstructed in the event that memory gets wiped for some reason. Theoretically, both the quad layout and the flex layout would get rebuilt the same way every time as long as the correct centerPosition and rotation values were used (and the layout code hadn't changed in the meantime). 

To recalculate centerPosition and rotation, we could just change initWithSpawn() to include all structures in the operation.staticStructures property. Even FlexLayout has a portion that is static that we can use to narrow it down to the exact position/rotation that was being used.

**Comment by bonzaiferroni** [comment]
I just thought of another idea, we could save the centerPosition and rotation to the controller sign, and look for that as well

**Comment by quonic** [comment]
The controller sign idea might work, but what about spawn position based on where the storage is located? Being that the storage is built long before a second spawn is built.

**Comment by bonzaiferroni** [comment]
Yeah, I'm tinkering with something right now that will look at all the core structures (roads, spawns, storages, etc.) and return the one that has the best match. Maybe the controller sign can be a fallback solution

**Comment by quonic** [comment]
Another way might to sort by the name of the spawns, if a user keeps the Spawn${i} format. Being that the lowest numbered spawn should be the first one built in that room.

**Comment by bonzaiferroni** [comment]
Alright, I just refactored findFromSpawn to recalculate the layout based on existing structures in the room, it is now called `ControllerOperation.recalculateLayout`. The old behavior should still work for people starting a brand new room, if it only finds one structure in the room it will first test for a compatible quad layout and fallback to a flex layout.

But if it finds more than one structure in the room, it will only test for the operation type specified by the flag, and it will look for the best match for centerPosition/rotation based on existing structures in the room.

Based on the testing that I did, it seems to work, here it has correctly found the parameters based on 3 spawns for flex layout:

https://screeps.slack.com/files/bonzai/F3DSLA89L/screenshot_from_2016-12-10_11-05-09.png

and here it has correctly found the paramaters for quad layout based on 2 spawns a storage:
https://screeps.slack.com/files/bonzai/F3D9A60TF/s70-recalclayout.gif
